drivers/retained_mem/retained_mem_nrf_gpregret.c

```diff
@@ -0,0 +1,147 @@
+{
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+#endif
+
+	return 0;
+}
+
+static int32_t nrf_gpregret_size(const struct device *dev)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+	return config->size;
+}
+
+static int32_t nrf_gpregret_read(const struct device *dev, size_t offset, uint8_t *buffer,
+				 size_t size)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy(buffer, (void *)(config->addr + offset), size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static int32_t nrf_gpregret_write(const struct device *dev, size_t offset, const uint8_t *buffer,
+				  size_t size)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy((void *)(config->addr + offset), buffer, size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static int32_t nrf_gpregret_clear(const struct device *dev)
+{
+	const struct nrf_gpregret_config *config = dev->config;
+
+#ifdef CONFIG_MULTITHREADING
+	struct nrf_gpregret_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memset((void *)config->addr, 0, config->size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
+
+	return 0;
+}
+
+static const struct retained_mem_driver_api nrf_gpregret_api = {
+	.size = nrf_gpregret_size,
+	.read = nrf_gpregret_read,
+	.write = nrf_gpregret_write,
+	.clear = nrf_gpregret_clear,
+};
+
+#define NRF_GPREGRET_DEVICE(inst)						\
+	static const struct nrf_gpregret_config nrf_gpregret_config_##inst = {	\
+		.addr = DT_INST_REG_ADDR(inst),					\
+		.size = DT_INST_REG_SIZE(inst),					\
+	};									\
+	DEVICE_DT_INST_DEFINE(inst,						\
+			      &nrf_gpregret_init,				\
+			      NULL,						\
+			      NULL,						\
+			      &nrf_gpregret_config_##inst,			\
+			      POST_KERNEL,					\
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,		\
+			      &nrf_gpregret_api);
+
+#define NRF_GPREGRET_DEVICE_MULTITHREADING(inst)				\
+	static struct nrf_gpregret_data nrf_gpregret_data_##inst;		\
+	static const struct nrf_gpregret_config nrf_gpregret_config_##inst = {	\
+		.addr = DT_INST_REG_ADDR(inst),					\
+		.size = DT_INST_REG_SIZE(inst),					\
+	};									\
+	DEVICE_DT_INST_DEFINE(inst,						\
+			      &nrf_gpregret_init,				\
+			      NULL,						\
+			      &nrf_gpregret_data_##inst,			\
+			      &nrf_gpregret_config_##inst,			\
+			      POST_KERNEL,					\
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,		\
+			      &nrf_gpregret_api);
+
+#ifdef CONFIG_MULTITHREADING
+DT_INST_FOREACH_STATUS_OKAY(NRF_GPREGRET_DEVICE_MULTITHREADING)
+#else
+DT_INST_FOREACH_STATUS_OKAY(NRF_GPREGRET_DEVICE)
+#endif
```

Comment:
Nonblocking: you could simplify this a bit by:

- moving the `nrf_gpregret_data` definition into an `IF_ENABLED(CONFIG_MULTITHREADING, ...)`
- using COND_CODE_1 to pass either data or NULL to DEVICE_DT_INST_DEFINE

 See spi_nrfx_spim.c for examples.

Reply:
Fixed